Miscellaneous arm64 vmm kernel changes

markjdb opened the pull request:
These address some of the issues I pointed out in the review, plus some additional bugs I found. If you're happy with the patches, please squash them into the main commit.

markjdb committed:
amd64/vmm: Remove unneeded include guards from vmm.h

vmm.h includes cpuset.h now, so they don't accomplish anything.

Sponsored by:	Innovate UK

markjdb committed:
arm64/vmm: Unconditionally unmap EL2 code and stacks when unloading

There is no particular reason to make this conditional on INVARIANTS,
and it's somewhat unusual for INVARIANTS-only code to have side effects.

Also clear additional pmap function pointers.

Sponsored by:	Innovate UK

markjdb committed:
arm64/vmm: Fix unmapping of hyp structures

We were previously failing to unmap hypctx structures, and we were using
the wrong size when unmapping the main hyp structure.  This resulted in
panics when unloading vmm.ko after having created a vmm.

Sponsored by:	Innovate UK

markjdb committed:
arm64/vmm: Remove an unneeded MALLOC_DECLARE()

Sponsored by:	Innovate UK

markjdb committed:
arm64/vmm: Use a uintptr_t for the stack top in arm_setup_vectors()

This lets us drop an unnecessary cast and is better for Morello.  No
functional change intended.

Sponsored by:	Innovate UK

zxombie commented on sys/arm64/vmm/vmm_arm64.c:
```diff
@@ -441,22 +441,23 @@ vmmops_modcleanup(void)
 
 	smp_rendezvous(NULL, arm_teardown_vectors, NULL, NULL);
 
-#ifdef INVARIANTS
```
Do we need to remove the stacks in the non-INVARIANTS case? It should only be needed to not hit an assert to check we've removed all entries. As we never call into the hypervisor after the above `smp_rendezvous` and are about to free all page tables removing the entries shouldn't be needed.

markjdb replied:
Hmm, I think you're right that this is not strictly necessary. But even so, why not just do the cleanup unconditionally? It seems quite strange to make this operation conditional on INVARIANTS, and it's fragile in the face of future changes to the vmmpmap_* interface.

zxombie replied:
It's only there to remove a false-positive in the memory leak `KASSERT` under `vmmpmap_fini`.

What future changes do you have planned that would make this needed?

markjdb replied:
I don't have any future changes planned, I just mean that the use of INVARIANTS here is making assumptions about the internals of vmmpmap_fini(). This is not how we generally use INVARIANTS elsewhere in the kernel, and this code is not performance-sensitive, so I still don't understand why we don't use the same code in INVARIANTS and non-INVARIANTS kernels.

If you insist I can drop this patch (and I do need to rework the commit message in any case), but this use of `#ifdef INVARIANTS` is wrong IMHO.

markjdb committed:
arm64/vmm: Remove a commented-out function call

This will be reconciled somehow when the amd64 and arm64 vmm
implementations are merged.

Sponsored by:	Innovate UK
